rsc/client/valve/replay.go

```diff
@@ -0,0 +1,92 @@
+package valve
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"strings"
+)
+
+var (
+	ErrDownloadRelpay = fmt.Errorf("failed to download the dota2 replay")
+)
+
+type ReplayDownloader interface {
+	// Download
+	// downloading the replay file is downloaded directly to the
+	// Valve replay proxy using the replayURL.
+	//
+	// replayURL contains the full raw URI to download the replay to the Valve replay proxy,
+	// for example http://replay153.valve.net/570/7132230434_1635105612.dem.bz2
+	// Based on that value, we can download the replay DotA2 file directly using that URL.
+	//
+	// destination parameter specifies the file destination to save the downloaded replay file.
+	Download(ctx context.Context, replayURL string, destination string) error
+}
+
+type replay struct {
+	httpClient *http.Client
+}
+
+func NewReplay(httpClient *http.Client) ReplayDownloader {
+	return &replay{
+		httpClient: httpClient,
+	}
+}
+
+func (r *replay) Download(ctx context.Context, replayURL string, destination string) error {
```

Comment:
would be good to check if we already have a replay file with the same id locally to avoid redownloading, either in this function or on the caller function later on.

Reply:
Mmmmm it will be good, so we not gonna download it twice right. Will be updated!

Reply:
updated, already add mechanism to check whether the file is already exist in the destination path or not